src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java

```diff
@@ -114,40 +136,142 @@ static SqlAuthenticationToken getSqlFedAuthTokenPrincipal(SqlFedAuthInfo fedAuth
+                if (logger.isLoggable(Level.FINEST)) {
+                    logger.finest(logContext + "Error loading PKCS12 certificate: " + e.getMessage());
+                }
+            }
+
+            if (clientApplication == null) {
+                // try loading X509 cert
+                X509Certificate cert = (X509Certificate) SQLServerCertificateUtils.loadCertificate(certFile);
+
+                if (logger.isLoggable(Level.FINEST)) {
+                    logger.finest(logContext + "certificate type: " + cert.getType());
+
+                    // we don't really need to do this, MSAL will fail if cert is not valid, but good to check here and throw with proper error message
+                    cert.checkValidity();
+                    logger.finest(logContext + "certificate: " + cert.toString());
+                }
+
+                PrivateKey privateKey = SQLServerCertificateUtils.loadPrivateKey(certKey, certKeyPassword);
+
+                IClientCredential credential = ClientCredentialFactory.createFromCertificate(privateKey, cert);
+                clientApplication = ConfidentialClientApplication.builder(aadPrincipalID, credential)
+                        .executorService(executorService)
+                        .setTokenCacheAccessAspect(PersistentTokenCacheAccessAspect.getInstance())
+                        .authority(fedAuthInfo.stsurl).build();
+            }
+
+            final CompletableFuture<IAuthenticationResult> future = clientApplication
+                    .acquireToken(ClientCredentialParameters.builder(scopes).build());
+            final IAuthenticationResult authenticationResult = future.get();
 
             if (logger.isLoggable(Level.FINEST)) {
-                logger.finest(logger.toString() + authenticationString + ": get FedAuth token integrated, realm name:"
-                        + kerberosPrincipal.getRealm());
+                logger.finest(
+                        logContext + authenticationResult.account() != null ? authenticationResult.account().username()
+                                + ": " : "" + ACCESS_TOKEN_EXPIRE + authenticationResult.expiresOnDate());
             }
 
+            return new SqlAuthenticationToken(authenticationResult.accessToken(), authenticationResult.expiresOnDate());
+        } catch (InterruptedException e) {
+            // re-interrupt thread
+            Thread.currentThread().interrupt();
+
+            throw new SQLServerException(e.getMessage(), e);
+        } catch (ExecutionException | IOException e) {
+            throw getCorrectedException(e, aadPrincipalID, authenticationString);
+        } catch (GeneralSecurityException e) {
+            // this includes all certificate exceptions
+            throw getCorrectedException(e, aadPrincipalID, authenticationString);
+        } catch (Exception e) {
+            throw getCorrectedException(e, aadPrincipalID, authenticationString);
```

Comment:
These catch blocks are all identical, can we just have the final block?
```suggestion
        } catch (Exception e) {
            throw getCorrectedException(e, aadPrincipalID, authenticationString);
```